### ISSUE TITLE
HH C++ Routing missing most of the turns info

### ISSUE THREAD

**Issue by mdykierek** [osmandapp/OsmAnd#19016]
### Description

I had 2 situations where OsmAnd+ shows only one, very distant direction while there were plenty of intersections that need guidance. One of them I describe here, the other one was https://www.openstreetmap.org/way/47460763 that showed only Head with straight arrow, while I was supposed to take A4 towards east.

### Steps to reproduce

Going from Mikulov in Czech Republic to Ruda Śląska in Poland.
The highway intersection in Brno has a closed link to D1 (https://www.openstreetmap.org/way/142312586).
The closure was after offline maps update in January 2024 so I had to take a detour.

### Actual result

After taking the detour, OsmAnd+ re-calculated the route and the only direction was to take slight right in a couple hundreds of kilometers - leave A4 in Ruda Śląska.

### Expected result

There should be plenty of guidance for many highway intersections e.g. Vyskov, Prostejov, Olomouc etc.

### Your Environment (required)

**WARNING** Crash-Logs **MAY** contain information you deem sensitive.
Review this **CAREFULLY** before posting your issue!

```
OsmAnd Version: OsmAnd+ 4.6.13
Android/iOS version: 13
Device model: Moto Edge 40 Pro
Crash-Logs: ?
```

**Comment by mdykierek** [comment]
Today I had the same misbehavior on a shorter route from Dworcowa 25 in Gliwice Poland to Wyzwolenia 11 in Ruda Śląska, Poland. I had to follow one way thus required route re-calculation. Then it skipped the A4 exit towards 925 and showed the left turn on the bridge. It means it didn't show the exit but the turn left that would come next. I stopped the navigation and started again. The exit appeared just fine.

**Comment by sonora** [comment]
It sounds like this may actually duplicate #18824.

**Comment by vshcherb** [comment]
I think it's related to omitting unnecessary turn announcements on highways. Thought it's not clear.
1. https://test.osmand.net/map/?start=45.803704,12.818105&finish=45.804804,12.900674&profile=car#18/45.79641/12.82641 (https://www.openstreetmap.org/way/47460763#map=15/45.7878/12.8351) Actually here as I see it shuold be displayed Go ahead (+C|TR). 
2.https://test.osmand.net/map/?start=49.157145,16.600288&finish=49.159137,16.604880&profile=car#16/49.1584/16.6009 - also works fine

**Comment by vshcherb** [comment]
This route:
- https://www.openstreetmap.org/directions?engine=fossgis_osrm_car&route=50.5159%2C18.3044%3B50.4884%2C17.5929#map=10/50.5574/17.9936
- https://test.osmand.net/map/?start=50.515900,18.304400&finish=50.488400,17.592900&profile=car#11/50.6076/17.7316

I see enough guidance, so I'm not sure what exactly turn is missing please highlight on the route

**Comment by vshcherb** [comment]
I'm sorry, I still not sure I understand what exactly information missing. In this issue - https://github.com/osmandapp/OsmAnd/issues/19065 - there is a specific example, so there is a link on commit what was done and why, at least I could understand that's a regression and 

I can't say without specific intersection is it good to have a lot of turn lanes on each intersection or not good. Probably it's a duplicate request for https://github.com/osmandapp/OsmAnd/issues/19065

**Comment by vshcherb** [comment]
Actually I have 1 suspect:
 C++ HH routing (Used by default) produces too few route info 
1. C++ https://test.osmand.net/map/?start=51.552119,7.207010&finish=51.820306,6.850721&profile=car&params=car,short_way,nativerouting,noglobalfile,calcmode:NORMAL,hhonly#11/51.6368/6.9437
2. Java https://test.osmand.net/map/?start=51.552119,7.207010&finish=51.820306,6.850721&profile=car&params=car,short_way,noglobalfile,calcmode:NORMAL,hhonly#12/51.6521/6.9621

**Comment by mdykierek** [comment]
I will try to write it as simply as I can.
During a REAL navigation I had to change the route due to a detour. OsmAnd had to adjust and re-calculate route because of that.
After the re-calculation, an over 200 km long route (from Brno) showed only a left turn almost at the destination. So a very significant number of intersections, highway exits were ignored. Even the last exit on A4 was ignored. 
When I canceled the navigation and started again, the guidance was back to normal.
There was nothing redundant. There were several exits under way, highway changes etc.
In case of Gliwice-Ruda Śląska route, there was a number of intersections in the city, several intersections, highway exits etc., but the scenario was the same. Only left turn after the exit from A4 towards north onto 1-Maja was shown.
The route on the map looked OK, only the directions were not present.

**Comment by vshcherb** [comment]
This is task is merged with unfinished task:
- Double check Gpx approximation
- Don't do makePrecise in C++
- Transfer code refactoring 
   - https://github.com/osmandapp/OsmAnd/commit/b84319bcfa8aff6fd0e68fc29cdbdcfdaf413e15
   - https://github.com/osmandapp/OsmAnd/commit/c2ab2b6a14736d91ee47463b3f92469320fbfdb4#diff-9bd83e19e51b4a256282b7c6ef8c15f7094081727807f6cfe537dca5aa17819cR886

**Comment by vshcherb** [comment]
I think for now this issue is fixed and ready to be tested

**Comment by mdykierek** [comment]
This needs to be reopen. I had exactly the same issue with 4.7.10 just today.
Going from Jaworze, Poland to Ruda Śląska, with re-calculation because I missed a turn in Bielsko-Biała S52->1. The only turn was the exit from A4 right to 925. Stopping the route and re-calculating "by hand" fixed the turn info.

**Comment by DmitryAlexei** [comment]
@mdykierek could you please provide coordinates of your starting and destination points? We need them to reproduce your issue. 
Or you can also create your route on the map https://osmand.net/map/ and post it here. Simply copy your route from the browser address line.

**Comment by mdykierek** [comment]
I do not think you can simulate the behavior as the route needs to be re-calculated because of missed turn automatically but anyway: https://osmand.net/map/?start=49.792172,18.935277&end=50.256065,18.865773&profile=car#9/50.1492/18.7592
The turn I missed: https://www.openstreetmap.org/way/24931388.296875
Then I ignored the detour recommended by OsmAnd (because it was much longer) and went this way: https://osmand.net/map/?start=49.792172,18.935277&end=50.256065,18.865773&via=49.844251,19.068453&profile=car#16/49.8424/19.0549
It kept giving proper turn info until I reached road no 1:https://www.openstreetmap.org/way/273487768.484375
after that the only turn info was regarding intersection (not even highway exit, but traffic lights) here: https://www.openstreetmap.org/way/168533691.734375.
About here: https://www.openstreetmap.org/way/243067680.265625 I stopped the navigation, set the destination again and it was OK until destination.
BOTTOM LINE IS: when route is recalculated because of deviation of already computed route, the turn info fails at some point. It was not immediate. It "realized" that I made a detour, and computed new route, then guided for a couple of manouvers, then forgot most of the intermediate points.

**Comment by mdykierek** [comment]
May I ask why this issue is still closed?

**Comment by vshcherb** [comment]
Because it's fixed and released

**Comment by vshcherb** [comment]
If you experience issue only during recalculation, then it's worth to open another issue cause the original issue was about all  possible turns in all situations
```
when route is recalculated because of deviation of already computed route, the turn info fails at some point. It was not immediate. It "realized" that I made a detour, and computed new route, then guided for a couple of manouvers, then forgot most of the intermediate points.
```

**Comment by mdykierek** [comment]
In my opinion the issue as described has not been fixed and it was about deviations/detour/re-calculation from the very beginning. If it was released, then I don't have access to this release. If a different issue was fixed e.g. "HH C++ something", then it should be closed rather than my report I guess.
I  am currently using 4.7.17 and experience it on a daily basis literally everyday.
If I were to report this issue again, I would have to repeat all the comments + this one I planned to report now:

1. It is my observation that all turns are missing until and including motorway exit. The manouver that is shown, is the first after highway. Like yesterday, I went from Ruda Śląska to Tarnowskie Góry and  the first turn info was on this roundabout exit: https://www.openstreetmap.org/way/1074165365
2. During the navigation I described above, I set the destination on my driveway that has access=private. Then I backed up and the route got re-calculated almost immediately after start. Pretty nasty considering that turn info was mostly gone.

**Comment by vshcherb** [comment]
Ok, could you please @DmitryAlexei reproduce it and create a separate issue

**Comment by mdykierek** [comment]
The case remains Closed. I don't see any mention of a new bug for this malfunction and I almost missed an exit just yesterday going from Ruda Śląska in Poland to Jablunkov in Czechia. I was supposed to go through Bielsko-Biała, but decided to go through Mikołów and Skoczów. Furthermore I took a shortcut to bypas Mikołów (https://www.openstreetmap.org/way/35477075).
This made OsmAnd re-calculate the route at least twice.
As the result turn info was lost again! The only guidance was for the roundabout in Cieszyn! (https://www.openstreetmap.org/way/87496227)
So all lane assist on intersections and traffic lights. No exit to S52. This slowly becomes a showstopper.
Should I make a new bug report and put link to this issue as it describes a lot of cases and the plausible reason.
